### PR TITLE
Initial commit for Slow game speed (07/27)

### PR DESCRIPTION
**Pull request by qqkookie**
New Feature: Slow game speed step
Added new slow game running speed step below normal speed.
'Slow Speed' is 1/4 of normal game speed.
Game can be paused/unpaused by pressing "Reduce/Increase game speed" shortcut keys.
Source clean up: Replaced all the speed step numbers  by enum constants.

slow-game-speed-step

### CODE REVIEW

**Comment by qqkookie** [comment]
I did the dreaded "rebase", although I still don't know what I should do by the rebase. Now commit log is as follows. Did I do the "rebase" correctly? 

$ git log
commit a4ef66ff562545f7b409d5b6acc8de7034bd3468 (HEAD -> slow-game, origin/slow-game)
Merge: 1036b96fc ad62e8ab1
Author: qqkookie <39845532+qqkookie@users.noreply.github.com>
Date:   Fri Jul 27 20:47:22 2018 +0900

    slow-game

commit 1036b96fc0d41ab58cf371c4faa45e7f324b7db9
Author: qqkookie <39845532+qqkookie@users.noreply.github.com>
Date:   Fri Jul 27 17:53:51 2018 +0900

    Source formatting

commit 02d2e41178072bda408f639afbc60173b0c1b90b
Author: qqkookie <39845532+qqkookie@users.noreply.github.com>
Date:   Fri Jul 27 17:29:59 2018 +0900

    Initial commit for Slow game speed (07/27)

commit e44e7bd0f4af60bac480c5da4c5137d7d9b9b8e9
Author: Hielke Morsink <Broxzier@users.noreply.github.com>
Date:   Wed Jul 25 22:49:20 2018 +0200

    Replace 0 and NULL with nullptr (#7827)

commit 3efbfa98f33f8eb2d0cf04de00c48cd638688fab
Author: OpenRCT2 git bot <gitbot@openrct2.org>

**Comment by Broxzier** [comment]
No... there are some merge commits and all these commits you didn't make shouldn't show up. Why did you rebase? There were no conflicts, right?

**Comment by qqkookie** [comment]
You said that I should do rebase before. Then What should I do now? Should I undo the rebase? 

I still don't know what rebase actually is, and why we need rebase before committing, i.e the purpose of doing rebase. I think that resolving conflict with OpenRCT2:develop branch should be enough.  I  guessed that it is prior requirement for commit. I am not sure I did the rebase correctly.

**Comment by Broxzier** [comment]
With a rebase you change your first commit's parent to a later commit (base). That way the history is made lineair (no merging) and your changes show up more clearly.


https://hades.github.io/2010/03/git-your-friend-not-foe-vol-4-rebasing/

**Comment by qqkookie** [comment]
Then, why not just let  the forked branch (qqkookie/OpenRCT2:slow-game in this case) fetch all the recent changes on upstream branch ( OpenRCT/OpenRCT2:develop in this case) after the initial fork, and make the current local source will conflict with  remote slow-game (updated to up-to-date with the fetch) branch. and resolve the conflicts by manually editing the conflicts. 
I think that is much cleaner way to resolve all the conflict with upstream branch. 
It can be done easily with diff/merge tools without resort to "rebase".

**Comment by qqkookie** [comment]
Rebasing is very wasteful procedure that generates lots of unnecessary and irrelevant intermediate conflicts. Let's assume that the forked branch did few commits during development phase.  During the development period there may be many commits on the upstream/develop branch. These commits on upstream had their own share of conflicts and  resolved laboriously.  

Now the develop and review phase of the forked branch has done. Now is time to rebase.  The objective  is to resolve all the conflicts with current HEAD of forked branch and HEAD of current upstream/develop branch. It can be done easily and quick with diff /merge tool. It just merges HEAD of forked branch and HEAD of upstream.  All the intermediate node/version/commits are irreverent. 
On the contrary, rebasing generates  lots of intermediate conflicts , all of them should be resolved correctly. It simply can generates several 1000's of conflict for long and complex feature development. 

To mitigates and reduce such  intermediate conflicts , we can delete intermediate commits during the rebasing procedure. But it simply negates the very purpose of using version control system like Git. Removing  intermediate commits should be used as last resort, not used routinely.

So I think that fetch up-to-date upstream changes/commits and resolving conflict as HEAD to HEAD diff is much, much better way to resolve source conflicts  than abusing "rebase" and deleting "intermediate commits" for purpose of easier rebasing. 

"intermediate commits" should not deleted or merged unless absolutely necessary.  It is the very spirit of the version control system, be it VCS, RCS, or Git.

**Comment by Broxzier** [comment]
You went from not knowing how to rebase to having a very detailed opinion about why rebasing is bad. It has worked well so far, it creates much more linear history which is easier to bisect, rebasing does not create conflicts - they were already there - and you solve them the same way you would solve it when merging branches. Nothing gets deleted by doing a rebase, unless you really mess something up while in the middle of a rebase.

> To mitigates and reduce such intermediate conflicts , we can delete intermediate commits during the rebasing procedure.

Eh? Don't do that, you're removing your history then, that only makes it worse.

In practise, especially for bigger PRs, you still need to make adjustments a few times before a PR gets merged. To keep yours up to date with develop, you keep merging develop into the feature branch. Like stated before, that's not nice and now how we do things.

The only two downsides I can see of rebasing vs merging branches is that you _may_ have to solve the same conflicts multiple times, but usually that only happens when you've worked on something for a long time with many commits. Then sometimes merging can be in favour. And the other is that the commit hash changes, but that's only an issue when referencing to them on GitHub or something.

For small PRs like this one, creating merge commits only makes the commits look messy. Usually you don't even have conflicts when rebasing on latest develop, and when you do

So in short: Just do the rebase when it's needed and properly, you may have to solve a few conflicts, but that's it. If you're not going to work with our workflow, then I'm afraid your PRs won't get merged.

**Comment by qqkookie** [comment]
But, in previous closed PR, jensj12 said that "It can be helpful to first squash all your own commits into one (with rebase HEAD~4 -i), that squashes out all intermediate versions, but removes history. When there's a conflict, the conflicted parts consists of two parts. ..."   ShellI reopen it for you?

So I thought that "squash" is recommended and routine procedure for rebasing. 
Sorry for  misunderstanding.

I have no intention of "teaching" better way of VCS. I am very real novice on Git.  This is first time use of Git. If my words were heard as "teaching", forgive me for my bad way of expression. English is not my native language. It is just naive reaction from my inexperience and ignorance on Git, and plain common sense. Sorry again for my "teaching".

**Comment by Broxzier** [comment]
squashing _is_ useful sometimes to make the history cleaner. For example for the line where you added a space and later removed it again. Squashing these commits together will show only the actual change and not the two whitespace changes. We usually squash small commits like these before merging into develop.

**Comment by qqkookie** [comment]
I don't object judicious and sparing use of squashing. Sometime, deleting commit is necessary evil, like undoing mistake.

But, jensj12 said that "It can be helpful to first squash **all your own commits into one** (with rebase HEAD~4 -i), that **squashes out all intermediate versions**, but removes history." (emphasis is mine)

So I thought that it is best practice or mandatory rule for rebase, here in OpenRCT2 to "squash" is all the all intermediate versions. I mistook because I didn't know what the "rebase" actually is or what purpose it try to serve. Sorry again for my misunderstanding and ignorance. Thank you for good and kind advice to correct me.

**Comment by qqkookie** [comment]
How can I resolve this error?
It says no file for :  "/home/travis/build/OpenRCT2/OpenRCT2/build/g2.dat".
But I did not changed any install or g2 resource handling related thing.

-- Installing: /home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/libopenrct2.so
-- Set runtime path of "/home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/libopenrct2.so" to "$ORIGIN"
-- Installing: /home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/openrct2
-- Set runtime path of "/home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/openrct2" to "$ORIGIN"
-- Installing: /home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/openrct2-cli
-- Set runtime path of "/home/travis/build/OpenRCT2/OpenRCT2/build/OpenRCT2/bin/openrct2-cli" to "$ORIGIN"
CMake Error at cmake_install.cmake:129 (file):
  file INSTALL cannot find
  "/home/travis/build/OpenRCT2/OpenRCT2/build/g2.dat".
FAILED: CMakeFiles/install.util 
cd /home/travis/build/OpenRCT2/OpenRCT2/build && /usr/bin/cmake -P cmake_install.cmake
ninja: build stopped: cannot make progress due to previous errors.

**Comment by qqkookie** [comment]
Now what should do next?

**Comment by Lastorder-DC** [comment]
Do not merge to update branch. use rebase instead. however, this should done in shell(not gui)

1. Clone your forked repo to somewhere
1. Add upstream repo(in this case, `OpenRCT2/OpenRCT2`) by executing `git remote add upstream https://github.com/OpenRCT2/OpenRCT2`
1. Fetch all upstream branch by executing `git fetch upstream`
1. Start rebasing by `git rebase upstream/develop`.
1. if there are no conflict, rebase is done! however, if there are any conflicts, you should solve it by yourself. to solve conflict, edit file manually and `git add <conflicted file>`, `git rebase --continue`. rebase should continued after manual edit.

**Comment by qqkookie** [comment]
Thank you @Lastorder-DC 
With kind advice and help of you,  rebasing  (` git rebase upstream/develop`) and push (`git push --force`) was done successfully. So what is next step to do?

**Comment by janisozaur** [comment]
> So what is next step to do?

You'll have to wait until someone  from the team tests and reviews your change.

There are still two other improvements to be done:
1. You'll have to amend the network version (a define in `network.cpp, just increase current value by one)
2. Clean up the history. Right now there are 17 commits, most of which subject does not indicate what is the reason for it. Please have a look at https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages

**Comment by ZehMatt** [comment]
I feel like having a time scale makes much more sense than constantly introducing new speed levels. I had this on my list of things I wanted to do eventually. So having a timescale of 1 is normal, 0.5 would run half the speed and 2.0 would run twice the speed, also given by that we could have buttons to go faster or slower by a simple increment/decrement of any value we like.

**Comment by qqkookie** [comment]
How to clean history? Did you mean listing by "git log"? Then, how can I clean history? 
What did you mean by word "clean"? Which one did you mean?
1. Remove most lengthy comments and shorten the explanation?
2. Add more and longer comments for better, with detailed explanation
3.  Or, should I delete all the intermediate commits and leave only single commit?

But @Broxzier told me not to do so:  From post above post:
>> To mitigates and reduce such intermediate conflicts , we can delete intermediate commits during the
>> rebasing procedure.
>  Eh? Don't do that, you're removing your history then, that only makes it worse.
> In practise, especially for bigger PRs, you still need to make adjustments a few times before a PR gets 
> merged. .

And please tell me the exact procedure  to "clean history", step by step, line by line. What command or tool  should I use to clean the history?

**Comment by Broxzier** [comment]
That last quote were two different things combined. The first part is a reaction to you saying you were removing commits. The second part is about why you should not use merge, but rebase for PRs.

> please tell me the exact procedure to "clean history", 

Just squash similar commits together, and explain why that change was necessary. Janisosaur linked a page that explains what a good commit should look like.

> step by step, line by line

You asked this before and that's just very tedious, especially if you end up doing something else. Just ask for help here or in the gitter chat when you get stuck on something. To squash and reorder commits, rebase with the `-i` switch is your friend.

**Comment by Broxzier** [comment]
@qqkookie I do like this feature, but I dislike the current implementation (mainly the arrays not matching the defines and speeds), and it needs to be rebased properly of course. Do you plan on finishing this PR? If you do, please go over the reviews and comments in your other PRs too to see what needs changing.
https://github.com/OpenRCT2/OpenRCT2/pull/7809
https://github.com/OpenRCT2/OpenRCT2/pull/7816

**Comment by qqkookie** [comment]
>  the arrays not matching the defines and speeds),
Did you mean this? 
<PRE>
static uint32_t GameSpeedSteps[GAMESPEED_HYPER + 1] = { 0, 25, 100, 200, 400, 1000, 10000 };
</PRE>
Then, how about this? GAMESPEED_HYPER + 1 =  7 = GAMESPEED_COUNT
static uint32_t GameSpeedSteps[GAMESPEED_COUNT] = { 0, 25, 100, 200, 400, 1000, 10000 };
Is it OK?

File: distribution/changelog.txt [hunk 1]
```diff
@@ -5,10 +5,12 @@
 - Feature: [#7694] Debug option to visualize paths that the game detects as wide.
 - Feature: [#7713] The virtual floor now takes land ownership rights into account.
 - Feature: [#7771] Danish translation.
+- Feature: [#7816] Slow game speed step.
```
Comment (wolfreak99) [inline distribution/changelog.txt]: Change to "[#7816, #7831] Slow game speed step.", move to end of list, and remove the duplicate lines

File: distribution/changelog.txt [hunk 2]
```diff
@@ -5,7 +5,6 @@
 - Feature: [#7694] Debug option to visualize paths that the game detects as wide.
 - Feature: [#7713] The virtual floor now takes land ownership rights into account.
 - Feature: [#7771] Danish translation.
-- Feature: [#7816] Slow game speed step.
```
Comment (wolfreak99) [inline distribution/changelog.txt]: You forgot to add the changelog back.

**Comment by Broxzier** [comment]
It looks like rebasing is not going right again. When rebasing, first fetch upstream, then rebase on top of a clean upstream/develop and pick only the commits you've made.

File: src/openrct2/Game.cpp [hunk 3]
```diff
@@ -129,17 +129,24 @@ GAME_COMMAND_CALLBACK_POINTER* game_command_callback_get_callback(uint32_t index
...
 {
-    gGameSpeed = std::min(gConfigGeneral.debugging_tools ? 5 : 4, gGameSpeed + 1);
-    if (gGameSpeed == 5)
-        gGameSpeed = 8;
+    gGameSpeed++;
+    if (gGameSpeed == GAMESPEED_HYPER && !gConfigGeneral.debugging_tools)
+        gGameSpeed--;
+
+    if (game_is_paused())
+        pause_toggle();
```
Comment (Broxzier) [inline src/openrct2/Game.cpp]: I'm not a fan of this. This means that the game speed cannot be changed while the game remains paused, since it will unpause right away. I'd leave the pause state to the player.

File: src/openrct2/Game.cpp [hunk 4]
```diff
@@ -822,6 +829,10 @@ void pause_toggle()
     {
         audio_stop_all_music_and_sounds();
     }
+    else if (gGameSpeed == GAMESPEED_PAUSED)
+    {
+        gGameSpeed = GAMESPEED_NORMAL;
```
Comment (Broxzier) [inline src/openrct2/Game.cpp]: Setting it to normal when unpausing feels odd when the speed was changed with hotkeys. When lowering it takes two clicks to go from normal to pause, but going back to normal only requires one click on the pause button.

I'm in favour of removing the pause logic from the game speed changes entirely, so that the lowest a player can go is "slow speed", and pausing is something else.

File: src/openrct2/GameState.cpp [hunk 5]
```diff
@@ -73,6 +76,7 @@ void GameState::InitAll(int32_t mapSize)
 
 void GameState::Update()
 {
+    static uint32_t pctUpdates = 0;
```
Comment (Broxzier) [inline src/openrct2/GameState.cpp]: Static variables won't work well when we want to support having multiple game-states in the future.

File: src/openrct2/GameState.cpp [hunk 6]
```diff
@@ -91,14 +95,21 @@ void GameState::Update()
     }
 
     // Determine how many times we need to update the game
-    if (gGameSpeed > 1)
+    if (gGameSpeed == GAMESPEED_NORMAL)
```
Comment (Broxzier) [inline src/openrct2/GameState.cpp]: Why does normal game speed need to be separated from the rest?
